generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -84,6 +85,134 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
+
+    static List<OpaqueZeroOneFactory> waysToMakeZero(OpaqueExpressionGenerator
+                                                         opaqueExpressionGenerator) {
+      return Arrays.asList(
+          opaqueExpressionGenerator::opaqueZeroOrOneFromIdentityFunction,
+          opaqueExpressionGenerator::opaqueZeroOrOneFromInjectionSwitch,
+          opaqueExpressionGenerator::opaqueZeroOrOneSquareRoot,
+          opaqueExpressionGenerator::opaqueZeroSin,
+          opaqueExpressionGenerator::opaqueZeroLogarithm
+      );
+    }
+
+    static List<OpaqueZeroOneFactory> waysToMakeOne(OpaqueExpressionGenerator
+                                                        opaqueExpressionGenerator) {
+      return Arrays.asList(
+          opaqueExpressionGenerator::opaqueZeroOrOneFromIdentityFunction,
+          opaqueExpressionGenerator::opaqueZeroOrOneFromInjectionSwitch,
+          opaqueExpressionGenerator::opaqueZeroOrOneSquareRoot,
+          opaqueExpressionGenerator::opaqueOneExponential,
+          opaqueExpressionGenerator::opaqueOneCosine,
+          opaqueExpressionGenerator::opaqueOneNormalizedVectorLength
+      );
+    }
+  }
+
+  private Optional opaqueZeroOrOneFromIdentityFunction(BasicType type, boolean constContext,
+                                                       final int depth,
+                                                       Fuzzer fuzzer, boolean isZero) {
+    // Make an opaque value recursively and apply an identity function to it
+    return Optional.of(applyIdentityFunction(makeOpaqueZeroOrOne(isZero, type, constContext,
+        depth, fuzzer),
+        type,
+        constContext,
+        depth,
+        fuzzer));
+  }
+
+  private Optional opaqueZeroOrOneFromInjectionSwitch(BasicType type, boolean constContext,
+                                                      final int depth,
+                                                      Fuzzer fuzzer, boolean isZero) {
+    // injectionSwitch.x or injectionSwitch.y
+    if (constContext || !generationParams.getInjectionSwitchIsAvailable()) {
+      return Optional.empty();
+    }
+    return Optional.of(makeOpaqueZeroOrOneFromInjectionSwitch(isZero, type));
+  }
+
+  private Optional opaqueZeroOrOneSquareRoot(BasicType type, boolean constContext, final int depth,
+                                             Fuzzer fuzzer, boolean isZero) {
+    // sqrt (opaque)
+    if (!BasicType.allGenTypes().contains(type)) {
+      return Optional.empty();
+    }
+    return Optional.of(new FunctionCallExpr("sqrt", makeOpaqueZeroOrOne(isZero, type, constContext,
+        depth, fuzzer)));
+  }
+
+  private Optional opaqueZeroSin(BasicType type, boolean constContext, final int depth,
+                                 Fuzzer fuzzer, boolean isZero) {
+    // represent 0 as sin(opaqueZero) function, e.g. sin(0.0)
+    if (!isZero) {
+      return Optional.empty();
+    }
+    if (!BasicType.allGenTypes().contains(type)) {
+      return Optional.empty();
```

Comment:
In contrast to my remark about the assertion above, it is indeed appropriate to return Optional.empty() here, since we may well call this method with an incompatible type.